File: backend/lib/backend/learning/task_attempt.ex
defmodule Backend.Learning.TaskAttempt do
  use Ecto.Schema
  import Ecto.Changeset

  alias Backend.Content.Task
  alias Backend.Learning.ChildProfile

  schema "task_attempts" do
    field :selected_answer, :string
    field :is_correct, :boolean, default: false
    field :hint_used, :boolean, default: false
    field :attempt_number, :integer

    belongs_to :child_profile, ChildProfile
    belongs_to :task, Task

    timestamps(type: :utc_datetime)
  end

  @doc false
  def changeset(task_attempt, attrs) do
    task_attempt
    |> cast(attrs, [
      :child_profile_id,
      :task_id,
      :selected_answer,
      :is_correct,
      :hint_used,
      :attempt_number
    ])
    |> validate_required([
      :child_profile_id,
      :task_id,
      :selected_answer,
      :is_correct,
      :hint_used,
      :attempt_number
    ])
    |> validate_number(:attempt_number, greater_than_or_equal_to: 1)
    |> foreign_key_constraint(:child_profile_id)
    |> foreign_key_constraint(:task_id)
  end
end

File: mobile/lib/models/task_attempt.dart
class TaskAttempt {
  const TaskAttempt({
    required this.id,
    required this.taskId,
    required this.selectedAnswer,
    required this.isCorrect,
    required this.attemptNumber,
    required this.hintUsed,
  });

  final int id;
  final int taskId;
  final String selectedAnswer;
  final bool isCorrect;
  final int attemptNumber;
  final bool hintUsed;

  factory TaskAttempt.fromJson(Map<String, dynamic> json) => TaskAttempt(
    id: (json['id'] as num?)?.toInt() ?? 0,
    taskId: (json['task_id'] as num?)?.toInt() ?? 0,
    selectedAnswer: json['selected_answer'] as String? ?? '',
    isCorrect: json['is_correct'] as bool? ?? false,
    attemptNumber: (json['attempt_number'] as num?)?.toInt() ?? 0,
    hintUsed: json['hint_used'] as bool? ?? false,
  );
}
